lib/output.js

```diff
@@ -69,50 +83,39 @@ function OutputStream(options) {
         quote_style      : 0,
         keep_quoted_props: false,
         wrap_iife        : false,
+        _comments        : null,
     }, true);
 
     // Convert comment option to RegExp if neccessary and set up comments filter
-    if (typeof options.comments === "string" && /^\/.*\/[a-zA-Z]*$/.test(options.comments)) {
-        var regex_pos = options.comments.lastIndexOf("/");
-        options.comments = new RegExp(
-            options.comments.substr(1, regex_pos - 1),
-            options.comments.substr(regex_pos + 1)
-        );
-    }
-    if (options.comments instanceof RegExp) {
-        options.comments = (function(f) {
-            return function(comment) {
-                return comment.type == "comment5" || f.test(comment.value);
-            }
-        })(options.comments);
-    }
-    else if (typeof options.comments === "function") {
-        options.comments = (function(f) {
-            return function(comment) {
-                return comment.type == "comment5" || f(this, comment);
-            }
-        })(options.comments);
-    }
-    else if (options.comments === "some") {
-        options.comments = function(comment) {
-            var text = comment.value;
-            var type = comment.type;
-            if (type == "comment2") {
-                // multiline comment
-                return /@preserve|@license|@cc_on/i.test(text);
-            }
-            return type == "comment5";
+    if (options.comments) {
+        var comments;
+        if (typeof options.comments === "string" && /^\/.*\/[a-zA-Z]*$/.test(options.comments)) {
+            var regex_pos = options.comments.lastIndexOf("/");
+            comments = new RegExp(
+                options.comments.substr(1, regex_pos - 1),
+                options.comments.substr(regex_pos + 1)
+            );
+        } else {
+            comments = options.comments;
         }
-    }
-    else if (options.comments){ // NOTE includes "all" option
-        options.comments = function() {
-            return true;
+        if (comments instanceof RegExp) {
+            options._comments = function(comment) {
+                return comment.type == "comment5" || comments.test(comment.value);
+            };
+        }
+        else if (typeof comments === "function") {
+            options._comments = function(comment) {
+                return comment.type == "comment5" || comments(this, comment);
+            };
+        }
+        else if (comments === "some") {
+            options._comments = some_comments;
+        } else { // NOTE includes "all" option
+            options._comments = return_true;
         }
     } else {
         // Falsy case, so reject all comments, except shebangs
-        options.comments = function(comment) {
-            return comment.type == "comment5";
-        }
+        options._comments = comment5;
```

Comment:
(uh nevermind, it's not the public interface, but I wouldn't touch it anyway for now)